### PR TITLE
Upload buildkite analytics .json files as a part of the upload jobs

### PR DESCRIPTION
**Pull request by staticfloat**
This re-uses the julia upload step to also do buildkite analytics
uploads.  This is already a signed job, and it already depends on the
tests running successfully, so it's a pretty natural re-use.

### CODE REVIEW

**Comment by vchuravy** [comment]
Thanks Elliot! I think my immediate thoughts are "Will we be able to map back to the steps" and if all of them are called results.json will we get all of them or only a single one.

**Comment by staticfloat** [comment]
> "Will we be able to map back to the steps"

That's what the `--step "test_${TRIPLET?}"` is for, so we only get a single `results.json` that maps to that exact step.

**Comment by vchuravy** [comment]
```
$ /etc/buildkite-agent/plugins/github-com-staticfloat-metahook-buildkite-plugin/hooks/pre-command
Missing artifact download path. See: `buildkite-agent artifact download --help`🚨 Error: The plugin metahook pre-command hook exited with status 1
```

File: pipelines/main/platforms/upload_macos.yml [hunk 1]
```diff
@@ -29,9 +29,19 @@ steps:
...
+                - BUILDKITE_ANALYTICS_TOKEN="U2FsdGVkX18BUSpNWNimRUOAh2ku0ugS7x3RRy1CfJBCxTLWArJpa1PXKoILfI7q"
               files:
                 - .buildkite/secrets/tarball_signing.gpg
                 - .buildkite/secrets/macos_codesigning.keychain
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "results.json"
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_macos.yml:41]:
```suggestion
                buildkite-agent artifact download --step "test_${TRIPLET?}" "results.json" .
```

File: pipelines/main/platforms/upload_linux.yml [hunk 2]
```diff
@@ -39,8 +39,19 @@ steps:
...
+                - BUILDKITE_ANALYTICS_TOKEN="U2FsdGVkX18BUSpNWNimRUOAh2ku0ugS7x3RRy1CfJBCxTLWArJpa1PXKoILfI7q"
               files:
                 - .buildkite/secrets/tarball_signing.gpg
+
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "results.json" .
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_linux.yml:51]:
```suggestion
                buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
```

File: pipelines/main/platforms/upload_windows.yml [hunk 3]
```diff
@@ -28,9 +28,19 @@ steps:
...
+                - BUILDKITE_ANALYTICS_TOKEN="U2FsdGVkX18BUSpNWNimRUOAh2ku0ugS7x3RRy1CfJBCxTLWArJpa1PXKoILfI7q"
               files:
                 - .buildkite/secrets/tarball_signing.gpg
                 - .buildkite/secrets/windows_codesigning.pfx
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "results.json"
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_windows.yml:40]:
```suggestion
                buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
```

File: pipelines/main/platforms/upload_freebsd.yml [hunk 4]
```diff
@@ -27,8 +27,18 @@ steps:
...
               files:
                 - .buildkite/secrets/tarball_signing.gpg
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
+          - test-collector#v1.10.1:
+              files: "results.json"
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_freebsd.yml:40]:
```suggestion
              files: "*/results.json"
```

File: pipelines/main/platforms/upload_linux.yml [hunk 5]
```diff
@@ -39,8 +39,19 @@ steps:
...
                 - .buildkite/secrets/tarball_signing.gpg
+
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
+          - test-collector#v1.10.1:
+              files: "results.json"
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_linux.yml:53]:
```suggestion
              files: "*/results.json"
```

File: pipelines/main/platforms/upload_macos.yml [hunk 6]
```diff
@@ -29,9 +29,19 @@ steps:
...
                 - .buildkite/secrets/tarball_signing.gpg
                 - .buildkite/secrets/macos_codesigning.keychain
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
+          - test-collector#v1.10.1:
+              files: "results.json"
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_macos.yml:43]:
```suggestion
              files: "*/results.json"
```

File: pipelines/main/platforms/upload_windows.yml [hunk 7]
```diff
@@ -28,9 +28,19 @@ steps:
...
                 - .buildkite/secrets/tarball_signing.gpg
                 - .buildkite/secrets/windows_codesigning.pfx
+          # In addition to uploading Julia, we will upload our buildkite analytics as well.
+          # Here we download the test `results.json` file and upload it with buildkite's `test-collector` plugin.
+          # This relies on the `BUILDKITE_ANALYTICS_TOKEN` secret unlocked by `cryptic` above.
+          - staticfloat/metahook:
+              pre-command: |
+                buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
+          - test-collector#v1.10.1:
+              files: "results.json"
```
Comment (vchuravy) [inline pipelines/main/platforms/upload_windows.yml:42]:
```suggestion
              files: "*/results.json"
```

File: pipelines/main/misc/upload_buildkite_results.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+          - staticfloat/cryptic#v2:
+              variables:
+                - BUILDKITE_ANALYTICS_TOKEN="U2FsdGVkX18BUSpNWNimRUOAh2ku0ugS7x3RRy1CfJBCxTLWArJpa1PXKoILfI7q"
+          - test-collector#v1.10.1:
+              files: "results.json"
+              format: "json"
+        timeout_in_minutes: ${TIMEOUT?}
+        soft_fail: true
+        commands: |
+          buildkite-agent artifact download --step "test_${TRIPLET?}" "*/results.json" .
```
Comment (vchuravy) [inline pipelines/main/misc/upload_buildkite_results.yml:22]:
```suggestion
          buildkite-agent artifact download --step "test_${TRIPLET?}" "results.json" .
```

We will see in a bit, but I think now the `/` will mess us up.

**Comment by staticfloat** [comment]
> {"message": "5660 results is greater than the maximum of 5000"}

🤦

**Comment by staticfloat** [comment]
Okay, [this is documented](https://buildkite.com/docs/test-analytics/importing-json#:~:text=prefixed%20fields%2C%20see%20CI%20environments.-,A%20single%20file%20can%20have%20a%20maximum%20of%205000%20test%20results%2C%20and%20if%20that%20limit%20is%20exceeded%20then%20the%20upload%20request%20will%20fail.%20To%20upload%20more%20than%205000%20test%20results%20for%20a%20single%20run%20upload%20multiple%20smaller%20files%20with%20the%20same%20run_env%5Bkey%5D.,-How%20to%20import%20JSON%20in%20CircleCI), we just need to split the `.json` files up I think.

**Comment by vchuravy** [comment]
Overall this looks good! I think the only failure case right now is test-ppc64 not producing any json and thus the upload step failing.

**Review by vchuravy — APPROVED** [review]

**Pull request merged**